presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregationIntoTableScan.java

```diff
@@ -0,0 +1,214 @@
+import io.prestosql.sql.planner.OrderingScheme;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.SymbolsExtractor;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.AggregationNode;
+import io.prestosql.sql.planner.plan.AggregationNode.GroupingSetDescriptor;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.plan.Patterns.aggregation;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.tableScan;
+
+public class PushAggregationIntoTableScan
+        implements Rule<AggregationNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .with(source().matching(tableScan().capturedAs(TABLE_SCAN)));
+
+    private final Metadata metadata;
+
+    public PushAggregationIntoTableScan(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode node, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+        Map<String, ColumnHandle> assignments = tableScan.getAssignments()
+                .entrySet().stream()
+                .collect(toImmutableMap(entry -> entry.getKey().getName(), Entry::getValue));
+
+        List<Entry<Symbol, AggregationNode.Aggregation>> aggregations = node.getAggregations()
+                .entrySet()
+                .stream()
+                .collect(toImmutableList());
+
+        List<AggregateFunction> aggregateFunctions = toAggregateFunction(aggregations
+                .stream()
+                .map(Entry::getValue)
+                .collect(toImmutableList()));
+        List<Symbol> aggregationOutputSymbols = aggregations.stream()
+                .map(Entry::getKey)
+                .collect(toImmutableList());
+
+        if (aggregateFunctions.isEmpty()) {
+            return Result.empty();
+        }
+
+        GroupingSetDescriptor groupingSets = node.getGroupingSets();
+
+        if (groupingSets.getGroupingSetCount() > 1) {
+            return Result.empty();
+        }
```

Comment:
Add this to the pattern:

```java
private static final Pattern<AggregationNode> PATTERN = aggregation()
        .matching(node -> node.getGroupingSets().getGroupingSetCount() == 1)
        .with(source().matching(tableScan().capturedAs(TABLE_SCAN)));
```

Reply:
Changed the pattern to capture this and no masks.